### PR TITLE
Refactor Kafka topics

### PR DESCRIPTION
**Pull request by disrupted**
aka "topics as first-class citizen"

### CODE REVIEW

**Review by sujuka99 — APPROVED** [review]

**Comment by sujuka99** [comment]
Please wait for @raminqaf for at least 1 more approval

File: kpops/components/streams_bootstrap/producer/model.py
Comment (raminqaf) [inline kpops/components/streams_bootstrap/producer/model.py:11]: Is this class needed? Can't we just on the field `KafkaStreamsConfig`?
```diff
- streams: ProducerStreamsConfig = Field(
        default=..., description=describe_attr("streams", __doc__)
    )
```
```diff
+ streams: KafkaStreamsConfig = Field(
        default=..., description=describe_attr("streams", __doc__)
    )
```

**Review by raminqaf — CHANGES_REQUESTED** [review]

File: kpops/components/base_components/models/topic.py
Comment (raminqaf) [inline kpops/components/base_components/models/topic.py:82]: Do we want the config to have a default value? Shouldn't we make this a mandatory field? What happens if the user doesn't define it and the default instance is used?

File: kpops/components/streams_bootstrap/producer/model.py
Comment (raminqaf) [reply kpops/components/streams_bootstrap/producer/model.py:11]: Sounds good! Just wanted to be sure this was the attended way!

File: kpops/component_handlers/topic/handler.py
Comment (disrupted) [reply kpops/component_handlers/topic/handler.py:31]: I didn't create this 😄 but you're right, I'll add some

File: kpops/component_handlers/kafka_connect/model.py
Comment (raminqaf) [inline kpops/component_handlers/kafka_connect/model.py:68]: I am wondering if this is the correct behaviour here. Shouldn't the "validate" method actually validate the input? For instance, if the user passes a malformed input. The method should check it and raise a ValidationError.

File: kpops/components/base_components/models/topic.py
Comment (raminqaf) [reply kpops/components/base_components/models/topic.py:82]: No need for it then. Maybe just add a small comment with the great explanation you just wrote! 😄

File: kpops/component_handlers/topic/handler.py
Comment (raminqaf) [inline kpops/component_handlers/topic/handler.py:137]: Not part of this PR but maybe we should look at the proxy wrapper and improve this logic. Back then we handled these with exception but maybe this would be cleaner (also more performant) with an if-else check.

Comment (disrupted) [reply kpops/component_handlers/topic/handler.py:137]: I called it `validate_topics` based on the usage of a Pydantic "`BeforeValidator`" and the examples in their docs. Essentially it is a custom deserializer (that's what the "before" indicates). 
from their docs:
> Before validators run before Pydantic's internal parsing and validation (e.g. coercion of a str to an int). These are more flexible than After validators since they can modify the raw input, but they also have to deal with the raw input, which in theory could be any arbitrary object.

The actual validation is done during by Pydantic during the init. If you want I can name it `deserialize_topics` instead

File: kpops/component_handlers/kafka_connect/model.py
Comment (disrupted) [reply kpops/component_handlers/kafka_connect/model.py:68]: I called it `validate_topics` based on the usage of a Pydantic "`BeforeValidator`" and the examples in their docs. Essentially it is a custom deserializer (that's what the "before" indicates). 
from their docs:
> Before validators run before Pydantic's internal parsing and validation (e.g. coercion of a str to an int). These are more flexible than After validators since they can modify the raw input, but they also have to deal with the raw input, which in theory could be any arbitrary object.

The actual validation is done during by Pydantic during the init. If you want I can name it `deserialize_topics` instead

File: kpops/component_handlers/topic/handler.py
Comment (disrupted) [reply kpops/component_handlers/topic/handler.py:137]: what do you have in mind to improve the performance here?
I agree that this (and most of our other handlers and API wrappers) needs refactoring, but I believe the exception is a good pattern in this case.

File: kpops/component_handlers/kafka_connect/model.py
Comment (raminqaf) [reply kpops/component_handlers/kafka_connect/model.py:68]: Yeah I like that name more. I was using this functionality in another project and at first it seemed very confusing to me!

File: kpops/component_handlers/topic/handler.py
Comment (raminqaf) [reply kpops/component_handlers/topic/handler.py:137]: I got mixed up with an article I read recently for Java. In Java this is considered as an anti-pattern. But apparently it's okay in Python. Here a short explanation:

https://stackoverflow.com/a/1835844

**Review by raminqaf — DISMISSED** [review]

File: tests/pipeline/snapshots/snap_test_example.py
Comment (raminqaf) [inline tests/pipeline/snapshots/snap_test_example.py:28]: Is this the correct behavior? The `generate` output does not show the `extraOutputTopics`, `from`, and `to` sections anymore.

Comment (disrupted) [reply tests/pipeline/snapshots/snap_test_example.py:28]: yep, extra output topics are hidden if empty now and cleaners shouldn't have `from` and `to` section

**Review by raminqaf — APPROVED** [review]

**Pull request merged**
